### PR TITLE
Login form: Add support for second-factor (form-based)

### PR DESCRIPTION
**Pull request by matthiaso**
354738

### CODE REVIEW

**Comment by stephan-merkli** [comment]
Change commit message to format [affected part]: [what was changed]
Login form: add support for second factor
or
Login form: add support for two-factor authentication

File: eclipse-scout-core/src/login/LoginBox.ts [hunk 1]
```diff
@@ -188,15 +223,18 @@ export class LoginBox extends Box {
...
-    this.$password
+    if (this.$password) {
+      this.$user
+        .val('')
+        .focus();
+      this.$user
+        .one('input.resetLoginError', this._resetButtonText.bind(this));
+    }
+
+    (this.$password ? this.$password : this.$token)
```
Comment (stephan-merkli) [inline eclipse-scout-core/src/login/LoginBox.ts]: Use local variable

File: eclipse-scout-core/src/login/LoginBox.ts [hunk 2]
```diff
@@ -188,15 +223,18 @@ export class LoginBox extends Box {
...
-      .val('');
 
-    this.$user
-      .one('input.resetLoginError', this._resetButtonText.bind(this));
-    this.$password
+    if (this.$password) {
+      this.$user
+        .val('')
+        .focus();
+      this.$user
```
Comment (stephan-merkli) [inline eclipse-scout-core/src/login/LoginBox.ts]: user should always be reset to. If second factor is invalid, login process should start at the begining again, not only with 2. factor (?)

File: org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java [hunk 3]
```diff
@@ -117,6 +124,69 @@ protected boolean handleAuthRequest(HttpServletRequest request, HttpServletRespo
...
+      return true;
+    }
+    int status = secondFactorVerifier.verify(firstFactorCompletedPrincipal.getName(), token.toCharArray());
+    if (status != ICredentialVerifier.AUTH_OK) {
+      handleForbidden(status, response);
+      return true;
+    }
+
+    // OWASP: force a new HTTP session to be created.
+    ServletFilterHelper helper = BEANS.get(ServletFilterHelper.class);
```
Comment (stephan-merkli) [inline org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java]: Create own method, similar code as in 117-123, add principal as parameter

File: org.eclipse.scout.rt.ui.html/src/main/resources/org/eclipse/scout/rt/ui/html/texts/Texts.properties [hunk 4]
```diff
@@ -161,6 +161,7 @@ ui.TableRowCount0=No row
 ui.TableRowCount1=One row
 ui.TableRowCount={0} rows
 ui.TileView=Tile view
+ui.Token=Token
```
Comment (stephan-merkli) [inline org.eclipse.scout.rt.ui.html/src/main/resources/org/eclipse/scout/rt/ui/html/texts/Texts.properties:164]: Required in UiTextContributor? Not sure, login box behaves differently.

File: org.eclipse.scout.rt.server.commons/pom.xml [hunk 5]
```diff
@@ -28,6 +28,10 @@
       <groupId>org.eclipse.scout.rt</groupId>
       <artifactId>org.eclipse.scout.rt.shared</artifactId>
     </dependency>
+    <dependency>
```
Comment (stephan-merkli) [inline org.eclipse.scout.rt.server.commons/pom.xml]: Would require Jackson dependency in .app too. Omit and create JSON manually for this simple case?

**Comment by matthiaso** [comment]
> Change commit message to format [affected part]: [what was changed] Login form: add support for second factor or Login form: add support for two-factor authentication

- [x] Changed?

File: org.eclipse.scout.rt.server.commons/pom.xml (hunk 5 above)
Comment (paolobazzi) [reply org.eclipse.scout.rt.server.commons/pom.xml]: Jackson dependency should be available on all .app modules.

rt.shared has already a (transitive) dependency to rt.dataobject through rt.security, therefore I suggest to remove this dependency here.

File: org.eclipse.scout.rt.ui.html/src/main/resources/org/eclipse/scout/rt/ui/html/texts/Texts.properties (hunk 4 above)
Comment (paolobazzi) [reply org.eclipse.scout.rt.ui.html/src/main/resources/org/eclipse/scout/rt/ui/html/texts/Texts.properties:164]: All text of scout.rt.ui.html are provided automatically to UI by org.eclipse.scout.rt.ui.html.UiTextContributor

File: eclipse-scout-core/src/login/LoginBox.ts (hunk 2 above)
Comment (paolobazzi) [reply eclipse-scout-core/src/login/LoginBox.ts]: Agree

File: org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java [hunk 6]
```diff
@@ -87,13 +91,11 @@ public void destroy() {
    */
   @SuppressWarnings({"squid:RedundantThrowsDeclarationCheck", "DuplicatedCode"}) // required so that overriding subclasses can throw ServletExceptions
   protected boolean handleAuthRequest(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java:93]: A unit test covering several cases for handleAuthRequest() would be nice ;-)
E.g. using mocked request and expectation on response?

**Review by paolobazzi — APPROVED** [review]
Looks good, see inputs for minor improvements.

File: eclipse-scout-core/src/login/LoginBox.ts (hunk 2 above)
Comment (matthiaso) [reply eclipse-scout-core/src/login/LoginBox.ts]: Also reset backend session

**Comment by matthiaso** [comment]
Changes since last review: https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..f580ed9a78b2af88bdf738a08859d722b6adfd5a

File: eclipse-scout-core/src/login/LoginBox.ts (hunk 2 above)
Comment (matthiaso) [reply eclipse-scout-core/src/login/LoginBox.ts]: Fixed in f580ed9a78b2af88bdf738a08859d722b6adfd5a

File: org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java (hunk 3 above)
Comment (matthiaso) [reply org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java]: See f580ed9a78b2af88bdf738a08859d722b6adfd5a, code is not so similar anymore as for two factor authentication session is now always reset. -> No own method created (yet).

File: eclipse-scout-core/src/login/LoginBox.ts (hunk 1 above)
Comment (matthiaso) [reply eclipse-scout-core/src/login/LoginBox.ts]: Code does not exist anymore (in favor of always resetting login completely).

**Comment by matthiaso** [comment]
Changes since last review: https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..c2fe96c0ff036a624eebaaee2060ce1777e1e8e1

File: org.eclipse.scout.rt.server.commons/pom.xml (hunk 5 above)
Comment (matthiaso) [reply org.eclipse.scout.rt.server.commons/pom.xml]: Removed dependency

File: org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java (hunk 6 above)
Comment (matthiaso) [reply org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java:93]: Fixed in c23066ab4cc592f8bef8ff6d062c5ab7fd48d8a0

**Comment by matthiaso** [comment]
Changes since last review https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..c23066ab4cc592f8bef8ff6d062c5ab7fd48d8a0

**Comment by matthiaso** [comment]
Still use https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..c23066ab4cc592f8bef8ff6d062c5ab7fd48d8a0 for comparision, last change only rebase

**Comment by matthiaso** [comment]
Still use https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..c23066ab4cc592f8bef8ff6d062c5ab7fd48d8a0 for comparision, last change only rebase

**Review by stephan-merkli — APPROVED** [review]

**Pull request merged**